src/Http/GraphRequest.php

```diff
@@ -394,163 +291,99 @@ function ($result) {
         }
-
-        return null;
     }
 
     /**
-    * Upload a file to OneDrive from a given location
-    *
-    * @param string $path   The path of the file to upload
-    * @param mixed  $client The client to use in the request
-    *
-    * @throws GraphException if file is invalid
-    * @throws \GuzzleHttp\Exception\GuzzleException
-    *
-    * @return mixed DriveItem or array of DriveItems
-    */
-    public function upload($path, $client = null)
+     * Upload a file from $path to Graph API
+     *
+     * @param string $path path of file to be uploaded
+     * @param ClientInterface|null $client (optional)
+     * @return array|GraphResponse|StreamInterface|object Graph Response object or response body cast to $returnType
+     * @throws ClientExceptionInterface|GraphClientException if $path cannot be opened for reading
+     */
+    public function upload(string $path, ?ClientInterface $client = null)
     {
         if (is_null($client)) {
-            $client = $this->createGuzzleClient();
+            $client = $this->graphClient->getHttpClient();
         }
         try {
-            if (file_exists($path) && is_readable($path)) {
-                $file = fopen($path, 'r');
-                $stream = \GuzzleHttp\Psr7\stream_for($file);
-                $this->requestBody = $stream;
-                return $this->execute($client);
-            } else {
-                throw new GraphException(GraphConstants::INVALID_FILE);
-            }
-        } catch(GraphException $e) {
-            throw new GraphException(GraphConstants::INVALID_FILE);
+            $resource = Utils::tryFopen($path, 'r');
+            $stream = Utils::streamFor($resource);
+            $this->attachBody($stream);
+            return $this->execute($client);
+        } catch(\RuntimeException $e) {
+            throw new GraphClientException(GraphConstants::INVALID_FILE, $e->getCode(), $e->getPrevious());
         }
     }
 
     /**
-    * Get a list of headers for the request
-    *
-    * @return array The headers for the request
-    */
-    private function _getDefaultHeaders()
-    {
-        $headers = [
-            'Host' => $this->baseUrl,
-            'Content-Type' => 'application/json',
-            'SdkVersion' => 'Graph-php-' . GraphConstants::SDK_VERSION,
-            'Authorization' => 'Bearer ' . $this->accessToken
-        ];
-        return $headers;
-    }
-
-    /**
-    * Get the concatenated request URL
-    *
-    * @return string request URL
-    */
-    private function _getRequestUrl()
+     * Sets default headers based on baseUrl being a Graph endpoint or not
+     */
+    private function initHeaders(string $baseUrl): void
     {
-        //Send request with opaque URL
-        if (stripos($this->endpoint, "http") === 0) {
-            return $this->endpoint;
+        $coreSdkVersion = "graph-php-core/".GraphConstants::SDK_VERSION;
+        $serviceLibSdkVersion = "Graph-php-".$this->graphClient->getSdkVersion();
```

Comment:
Should we change the service library SDK version header to match https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/middleware/TelemetryHandler.md#requirements?

Might need some changes to our PowerBI usage charts